File: packaging/src/main/javascript/wa/components/backup/Backup.js
/*
 * Copyright (c) 2002-2011 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * A component that handles setting up and triggering online backups.
 * 
 * Init module.
 */
wa.components.backup.Backup = (function($) {
    
    var me = {};
    
    me.basePage = $("<div></div>");
    me.ui = {};
    
    me.uiLoaded  = false;
    me.serverChanged = false;
    me.schedule = null;
    
    me.currentBackupPath = "";
    
    me.prevAction = null;
    
    me.visible = false;
    
    //
    // PUBLIC
    //
    
    me.api = {
            
            getPage :  function() {
                return me.basePage;
            },
            
            pageChanged : function(ev) {
                
                if(ev.data === "backup") {
                    
                    me.visible = true;
                    
                    if( me.uiLoaded === false ) {
                        me.uiLoaded = true;
                        me.basePage.setTemplateURL("templates/components/backup/index.tp");
                        me.render();
                    	me.loadBackupData();
                    } else if( me.serverChanged ) {
                    	me.serverChanged = false;
                        me.render();
                    	me.loadBackupData();
                    }
                    
                } else {
                    me.visible = false;
                }
            },
            
            serverChanged : function(ev) {
                
                me.currentBackupPath = "";
                
                if( me.visible === true ) {

                	// Load current backup path
                	me.loadBackupData();
                    me.render();
                    
                } else {
                	me.serverChanged = true;
                }
                
            },
            
            init : function() {
                me.loadBackupData();
            }
            
    };
    
    // 
    // PRIVATE
    //
    
    function serverManager() {
        return wa.Servers.getCurrentServer().manage;
    }
    
    me.render = function() {
    	
        me.basePage.processTemplate({
            server : me.server
        });
        
        delete(me.ui.jobList);
    };
    
    me.loadBackupData = function() {
    	
        var server = wa.Servers.getCurrentServer();
        
        if( server ) { 
        	server.manage.config.getProperty("general.backup.path", function(prop) {
        		me.currentBackupPath = prop.value;
        		me.updateUiBackupPath();
        	});
        	
        	updateJobs();
        }
    };
    
    me.updateUiBackupPath = function() {
    	$('.mor_backup_path').val(me.currentBackupPath);
		$('button.mor_backup_setpathbutton').attr('disabled', 'disabled');
		
		me.hideStatus();
		
		if(me.currentBackupPath.length > 0) {
			$('button.mor_backup_triggerbutton').removeAttr('disabled');
		} else {
			$('button.mor_backup_triggerbutton').attr('disabled', 'disabled');
		}
    };
    
    me.updateBackupJobUi = function(data) {
    	
    	if( ! me.ui.jobList ) {
    		me.ui.jobList = $("ul.mor_backup_job_list");
    		me.ui.jobList.setTemplateURL("templates/components/backup/schedule.tp");
    	}
    	
    	var data = wa.components.backup.parseJobData(data);
    	
    	me.ui.jobList.processTemplate({
    		jobs : data.jobList
    	});
    	
    };
    
    /**
     * Show a simple status message. This will hide the big backup
     * foundation-message.
     */
    me.showStatus = function(message) {
    	$('p.mor_backup_status').html(message);
		$('p.mor_backup_status').show();
		$('div.mor_backup_foundationbox').hide();
    };
    
    /**
     * Hide all status messages.
     */
    me.hideStatus = function() {
    	$('p.mor_backup_status').hide();
		$('div.mor_backup_foundationbox').hide();
    };

    function updateJobs() {
        if( me.visible ) {
            var server = wa.Servers.getCurrentServer();
            if(server && server.manage.backup.available) {
                server.manage.backup.getJobs(me.updateBackupJobUi);
            }
        }
    }
    
    //
    // CONSTRUCT
    //
    
    $('input.mor_backup_path').live('keyup',function(ev) {
    	 var el = $(ev.target);
    	 if( el.val().trim() === me.currentBackupPath ) {
    		 $('button.mor_backup_setpathbutton').attr('disabled', 'disabled');
    		 
    		 if( me.currentBackupPath.length > 0) {
    			 $('button.mor_backup_triggerbutton').removeAttr('disabled');
    		 }
    		 
    	 } else {
    		 $('button.mor_backup_setpathbutton').removeAttr('disabled');
    		 $('button.mor_backup_triggerbutton').attr('disabled', 'disabled');
    	 }
    });
    
    $('button.mor_backup_setpathbutton').live('click',function(ev) {
    	$('input.mor_backup_path').attr('disabled', 'disabled');
    	
    	var server = wa.Servers.getCurrentServer();
        
        server.manage.config.setProperty("general.backup.path", $('input.mor_backup_path').val(), function(result) {
            $('input.mor_backup_path').removeAttr('disabled');
            
            me.currentBackupPath = $('input.mor_backup_path').val();
            me.updateUiBackupPath();
            
        });
    	
    });
    
    $('button.mor_backup_triggerbutton').live('click', function(ev) {
    	
        var server = wa.Servers.getCurrentServer();
        
        me.showStatus("Performing backup..");
        $('button.mor_backup_triggerbutton').attr('disabled', 'disabled');
        
        server.manage.backup.triggerManual(function(data) {
            if(data !== false) {
                me.showStatus("Backup successful!");
                $('button.mor_backup_triggerbutton').removeAttr('disabled');
            } else {
                $('p.mor_backup_status').hide();
                $('button.mor_backup_triggerbutton').attr('disabled', 'disabled');
                $('div.mor_backup_foundationbox').show();
            }
        });
        
    });
    
    $('button.mor_backup_foundation_triggerbutton').live('click', function(ev) {
    	
    	if( confirm("This will DESTROY any files in '" + me.currentBackupPath + "', are you sure?") ) {
	    	
    	    var server = wa.Servers.getCurrentServer();
    	    
    	    me.showStatus("Creating foundation..");
            
            server.manage.backup.triggerManualFoundation(function(data) {
                me.showStatus("Foundation successful! Your database is now ready for online backups.");
                $('button.mor_backup_triggerbutton').removeAttr('disabled');            
                
            });
    	}
    	
    });
    
    $('button.mor_backup_add_job').live('click',function(ev) {
    	
    	wa.ui.Dialog.showUsingTemplate("New backup job","templates/components/backup/job.tp");
    	
    });
    
    $('button.mor_job_dialog_save').live('click', function(ev){
    	
    	// Validate form
    	var validators = [
		    { field:'input.mor_job_dialog_name',
		      validator : 'not_empty' },
		    { field:'input.mor_job_dialog_path',
	    	  validator : 'not_empty' },
	    	{ field:'input.mor_job_dialog_cronexp',
	    	  validator : 'not_empty' }
    	];
    	
    	
    	if( wa.FormValidator.validateFields(validators) ) {
    		
    		var name = $('input.mor_job_dialog_name').val();
			var path = $('input.mor_job_dialog_path').val();
			var cron = $('input.mor_job_dialog_cronexp').val();
			var autoFoundation = $('input.mor_job_dialog_auto-foundation:checked').length > 0;
    		
    		var id = $("input.mor_job_dialog_id").val();
    		
    		id = id.length > 0 ? id : null;
    		
    		// Save job
    		
    		var server = wa.Servers.getCurrentServer();
    		
    		server.manage.backup.setJob(
    		    {
    		        'name': name,
    		        'backupPath':path,
    		        'cronExpression':cron,
    		        'autoFoundation': autoFoundation,
    		        'id':id
    		    },
				function(){
    		        server.manage.backup.getJobs(me.updateBackupJobUi);
				}
    		);
    		
    		wa.ui.Dialog.close();
    	}
    	
    });
    
    $('button.mor_backup_job_edit').live('click', function(ev){
        var server = wa.Servers.getCurrentServer();
    	server.manage.backup.getJob($(ev.target).closest("li.mor_backup_job").find(".mor_backup_job_id_value").val(), function(job){
    	    neo4j.log(job);
    		wa.ui.Dialog.showUsingTemplate("Edit backup job","components/morpheus.backup/templates/job.tp", job);
    	});
    });
    
    
    $('button.mor_backup_job_delete').live('click', function(ev){
    	if( confirm("Are you sure you want to delete the backup job?") ) {
    	    var server = wa.Servers.getCurrentServer();
            server.manage.backup.deleteJob($(ev.target).closest("li.mor_backup_job").find(".mor_backup_job_id_value").val(), updateJobs);
    	}
    });
    
    $('button.mor_backup_job_create_foundation').live('click', function(ev){
    	var id = $(ev.target).closest("li.mor_backup_job").find(".mor_backup_job_id_value").val();
    	var server = wa.Servers.getCurrentServer();
        server.manage.backup.getJob(id, function(job){
    		if(job && confirm("This will delete any current backup in " + job.backupPath +". Are you sure?")) {
    			
    			$(".mor_backup_job_info").html("Creating foundation..");
    			$(".mor_backup_job_info").show();
    			
    			server.manage.backup.triggerJobFoundation(job.id, 
    				function(data) {
    					$(".mor_backup_job_info").html("Successfully created foundation.");
    					setTimeout(function() {
    						$(".mor_backup_job_info").hide();
    					}, 2000);
    				});
    			
    		}
    	});
    });
    
    // Update job info at regular intervals
    
    setInterval(updateJobs, 5000);
    
    return me.api;
    
})(jQuery);

//
// REGISTER STUFF
//

wa.ui.Pages.add("backup",wa.components.backup.Backup);
wa.ui.MainMenu.add({ label : "Backup", pageKey:"backup", index:4, requiredServices:['backup'], perspectives:['server']});

wa.bind("init", wa.components.backup.Backup.init);
wa.bind("ui.page.changed", wa.components.backup.Backup.pageChanged);
wa.bind("servers.current.changed", wa.components.backup.Backup.serverChanged);

File: enterprise/src/main/java/org/neo4j/backup/Backup.java
/**
 * Copyright (c) 2002-2011 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.neo4j.backup;

import org.apache.zookeeper.KeeperException;
import org.neo4j.com.ComException;
import org.neo4j.com.backup.OnlineBackup;
import org.neo4j.helpers.Args;
import org.neo4j.helpers.Pair;
import org.neo4j.kernel.ha.zookeeper.ClusterManager;
import org.neo4j.kernel.ha.zookeeper.Machine;

public class Backup
{
    private static final String TO = "to";
    private static final String FROM_HA = "from-ha";
    private static final String FROM = "from";
    private static final String INCREMENTAL = "incremental";
    private static final String FULL = "full";

    public static void main( String[] args )
    {
        Args arguments = new Args( args );
        boolean full = arguments.has( FULL );
        boolean incremental = arguments.has( INCREMENTAL );
        if ( full&incremental || !(full|incremental) )
        {
            System.out.println( "Specify either " + dash( FULL ) + " or " + dash( INCREMENTAL ) );
            exitAbnormally();
        }
        
        String from = arguments.get( FROM, null );
        String fromHa = arguments.get( FROM_HA, null );
        if ( (from != null && fromHa != null) || (from == null && fromHa == null) )
        {
            System.out.println( "Specify either " + dash( FROM ) + " or " + dash( FROM_HA ) );
            exitAbnormally();
        }
        
        String to = arguments.get( TO, null );
        if ( to == null )
        {
            System.out.println( "Specify target location with " + dash( TO ) + " <target-directory>" );
            exitAbnormally();
        }
        
        if ( fromHa != null )
        {
            // This means we're trying to reach an HA cluster, a ZooKeeper service
            // managing that cluster, that is.
            try
            {
                System.out.println( "Asking ZooKeeper service at '" + fromHa + "' for master" );
                from = getMasterServerInCluster( fromHa );
                System.out.println( "Found master '" + from + "' in cluster" );
            }
            catch ( ComException e )
            {
                System.out.println( e.getMessage() );
                exitAbnormally();
            }
            catch ( RuntimeException e )
            {
                if ( e.getCause() instanceof KeeperException )
                {
                    KeeperException zkException = (KeeperException) e.getCause();
                    System.out.println( "Couldn't connect to '" + fromHa + "', " + zkException.getMessage() );
                    exitAbnormally();
                }
                throw e;
            }
        }
        
        doBackup( full, from, to );
    }

    private static void doBackup( boolean trueForFullFalseForIncremental, String from, String to )
    {
        OnlineBackup backup = newOnlineBackup( from );
        try
        {
            if ( trueForFullFalseForIncremental )
            {
                System.out.println( "Performing full backup from '" + from + "'" );
                backup.full( to );
            }
            else
            {
                System.out.println( "Performing incremental backup from '" + from + "'" );
                backup.incremental( to );
            }
            System.out.println( "Done" );
        }
        catch ( ComException e )
        {
            System.out.println( "Couldn't connect to '" + from + "', " + e.getMessage() );
            exitAbnormally();
        }
    }

    private static void exitAbnormally()
    {
        System.exit( 1 );
    }

    private static String dash( String name )
    {
        return "-" + name;
    }

    private static OnlineBackup newOnlineBackup( String from )
    {
        if ( from.contains( ":" ) )
        {
            int colonIndex = from.indexOf( ':' );
            String host = from.substring( 0, colonIndex );
            int port = Integer.parseInt( from.substring( colonIndex+1 ) );
            return OnlineBackup.from( host, port );
        }
        return OnlineBackup.from( from );
    }

    private static String getMasterServerInCluster( String from )
    {
        ClusterManager clusterManager = new ClusterManager( from );
        Pair<String, Integer> masterServer = null;
        try
        {
            clusterManager.waitForSyncConnected();
            Machine master = clusterManager.getMaster();
            masterServer = master.getServer();
            if ( masterServer != null )
            {
                int backupPort = clusterManager.getBackupPort( master.getMachineId() );
                return masterServer.first() + ":" + backupPort;
            }
            throw new ComException( "Master couldn't be found from cluster managed by " + from );
        }
        finally
        {
            clusterManager.shutdown();
        }
    }
}
